sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala

```diff
@@ -206,6 +206,18 @@ trait CodegenSupport extends SparkPlan {
   def doConsume(ctx: CodegenContext, input: Seq[ExprCode], row: ExprCode): String = {
     throw new UnsupportedOperationException
   }
+
+  /**
+   * for optimization to suppress shouldStop() in a loop of WholeStageCodegen
+   *
+   * isShouldStopRequired: require to insert shouldStop() into the loop if true
+   */
+  def isShouldStopRequired: Boolean = {
+    shouldStopRequired || (this.parent != null && this.parent.isShouldStopRequired)
+  }
+
+  // set true if doConsume() inserts append() method that requires shouldStop() in the loop
+  protected var shouldStopRequired: Boolean = false
```

Comment:
Do you think it is possible to do this without a mutable var? Code generation has way to much mutable state as it is.